controllers/complianceeventsapi/auth.go

```diff
@@ -0,0 +1,165 @@
+package complianceeventsapi
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/stolostron/rbac-api-utils/pkg/rbac"
+	"gopkg.in/square/go-jose.v2/jwt"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	policyv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+)
+
+func getReadAuth(ctx context.Context, config *rest.Config,
+	token, managedClusterName string, client client.Client,
+) (bool, error) {
+	allManagedClusterRules, err := getAllManagedClusterRules(ctx, config, token, client)
+	if err != nil {
+		return false, err
+	}
+
+	rules, ok := allManagedClusterRules[managedClusterName]
+	if !ok {
+		return false, nil
+	}
+
+	if slices.Contains(rules, "get") || slices.Contains(rules, "*") {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func getAllManagedClusterRules(ctx context.Context, config *rest.Config,
+	token string, client client.Client,
+) (map[string][]string, error) {
+	mangedClusterList := &clusterv1.ManagedClusterList{}
+
+	err := client.List(ctx, mangedClusterList)
+	if err != nil {
+		return nil, err
+	}
+
+	managedClusterNames := make([]string, 0, len(mangedClusterList.Items))
+
+	for _, mc := range mangedClusterList.Items {
+		managedClusterNames = append(managedClusterNames, mc.Name)
+	}
+
+	return getManagedClusterRules(config,
+		token, managedClusterNames)
+}
+
+func getManagedClusterRules(config *rest.Config,
+	token string, managedClusterNames []string,
+) (map[string][]string, error) {
+	saName := getServiceAccountName(token)
+	if saName == "" {
+		return nil, fmt.Errorf("failed to parse token")
+	}
+	// Token include Bearer prefix
+	config.Impersonate = rest.ImpersonationConfig{
```

Comment:
We don't want to use impersonation because then group membership does not get taken into account.

We want to instead use the user token to make an SelfSubjectAccessReview.

For example, see how the Kubernetes client is generated using the input token:
https://github.com/stolostron/rbac-api-utils/blob/159deac7d398c3470ae730c3fbb8ffbbcca1ac5a/pkg/rbac/rbac.go#L91-L104

Then you can call `GetResourceAccess` with the Kubernetes client.

Reply:
Thanks!